test/distributed/test_ddp_under_dist_autograd.py

```diff
@@ -0,0 +1,410 @@
+    def _master_process(self, ddp_mode: DdpMode):
+        gLogger.info(f"Running the master process...")
+        process_group_for_ddp = dist_c10d.new_group(ranks=TRAINER_RANKS)
+        remote_em_rref = rpc.remote(
+            self.remote_worker_name(), RemoteEM, args=(NUM_EM_ROW, D_SPARSE)
+        )
+        remote_net_rref = rpc.remote(
+            self.remote_worker_name(),
+            RemoteNet,
+            args=(D_DENSE + D_SPARSE, D_HID),
+        )
+        gLogger.info(f"Created remote rrefs on master")
+        self.do_test_on_master(ddp_mode, remote_em_rref, remote_net_rref)
+        dist.destroy_process_group(process_group_for_ddp)
+
+    def do_test_on_master(
+        self,
+        ddp_mode: DdpMode,
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+    ):
+        trainer_rrefs = []
+        for rank in TRAINER_RANKS:
+            trainer = self.trainer_name(rank)
+            trainer_rrefs.append(
+                rpc.remote(
+                    trainer,
+                    Trainer,
+                    args=(remote_em_rref, remote_net_rref, ddp_mode, rank),
+                )
+            )
+
+        training_examples = get_training_examples()
+        for _ in range(3):
+            futures = []
+            for idx, trainer_rref in enumerate(trainer_rrefs):
+                futures.append(
+                    _remote_method_async(
+                        Trainer.do_backward,
+                        trainer_rref,
+                        training_examples[idx],
+                    )
+                )
+
+            for future in futures:
+                ddp_grads, non_ddp_grads = future.wait()
+                for grad in ddp_grads:
+                    self.assertAlmostEqual(
+                        0.0,
+                        float(grad.norm()),
+                        msg="The grad for any ddp parameter should be zeros, because "
+                        "the training examples' grads cancel each other.",
+                    )
+                for grad in non_ddp_grads:
+                    self.assertNotAlmostEqual(
+                        0.0,
+                        float(grad.norm()),
+                        msg="The grad for any non-ddp parameter shouldn't be zeros",
+                    )
+
+    def _do_test(self, ddp_mode):
+        if self.rank == MASTER_RANK:
+            self._master_process(ddp_mode)
+        elif self.rank == REMOTE_WORKER_RANK:
+            self._remote_worker_process()
+        elif self.rank in TRAINER_RANKS:
+            self._trainer_process(self.rank)
+        else:
+            raise RuntimeError(f"Unknow process rank: {self.rank}")
+
+    @requires_gloo()
+    def test_backward_no_ddp(self):
+        self._do_test(DdpMode.NONE)
+
+    @requires_gloo()
+    def test_backward_ddp_outside(self):
+        self._do_test(DdpMode.OUTSIDE)
+
+    @requires_gloo()
+    def test_backward_ddp_inside(self):
+        self._do_test(DdpMode.INSIDE)
```

Comment:
I think we need the following additional unit tests:

1) Use a model which doesn't have any RPC, run its forward pass and then run the backward pass using both local engine and distributed engine and verify that the results are the same.
2) Run the forward and backward pass for a model which has RPC and DDP with distributed autograd. Now run a similar model where the RPC is replaced with a local op (which is equivalent) and run this model with local autograd. Compare the results of both of these to ensure correctness. We do something similar for our distributed autograd tests where we compare its results to the local autograd engine: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/rpc/dist_autograd_test.py#L932.

In addition to this, I think the entire test suite needs to be also run for NCCL where the DDP params are on GPUs.

Reply:
> In addition to this, I think the entire test suite needs to be also run for NCCL where the DDP params are on GPUs.

Is there an existing test case for DDP on NCCL I can learn from?

Reply:
@pritamdamania87 
I added a test case to compare local and distributed backward runs on a DDP module without RPC.

The other 2 test cases are quite complicated. I'd prefer adding them in follow-up PRs. What do you think?